### PR TITLE
use native long instead of Long wrapper for ShallowTrace

### PR DESCRIPTION
**Pull request by aman1309**
Solves https://github.com/linkedin/parseq/issues/244

### CODE REVIEW

**Comment by aman1309** [comment]
> Did we see any performance improvement in running benchmark profiles with this change?

There is a slight increase tried 3 times on Mac with similar conditions
Before:
Throughput: 999.362 plans/s, 300807.976 tasks/s

After:
Throughput: 1001.011 plans/s, 301304.197 tasks/s

**Comment by mchen07** [comment]
> > Did we see any performance improvement in running benchmark profiles with this change?
> 
> There is a slight increase tried 3 times on Mac with similar conditions
> Before:
> Throughput: 999.362 plans/s, 300807.976 tasks/s
> 
> After:
> Throughput: 1001.011 plans/s, 301304.197 tasks/s

The initial motivation for this change is memory over-use from our ShallowTrace data structure, is it possible to measure the memory usage before and after?

File: subprojects/parseq/src/test/java/com/linkedin/parseq/TestTaskStates.java [hunk 1]
```diff
@@ -268,8 +268,8 @@ private void runTask(final Task<?> task) {
   private void assertInitOrScheduled(final Task<?> task) {
     assertFalse(task.isDone());
     assertFalse(task.isFailed());
-    assertNull(task.getShallowTrace().getStartNanos());
-    assertNull(task.getShallowTrace().getEndNanos());
+    assertEquals(task.getShallowTrace().getNativeStartNanos(), -1);
```
Comment (mchen07) [inline subprojects/parseq/src/test/java/com/linkedin/parseq/TestTaskStates.java:271]: We should create new tests for new APIs to make sure that our change will not break existing users.

**Comment by junchuanwang** [comment]
+1 for memory footprint. Can we do a heap dump?

File: subprojects/parseq/src/test/java/com/linkedin/parseq/TestTaskStates.java (hunk 1 above)
Comment (aman1309) [reply subprojects/parseq/src/test/java/com/linkedin/parseq/TestTaskStates.java:271]: to keep it clean, I had tested these Uts with deprecated methods. added new tests for deprecated methods separately. Check TestShallowTraceBuilder#testDeprecatedTimeMethods

**Comment by aman1309** [comment]
> > > Did we see any performance improvement in running benchmark profiles with this change?
> > 
> > 
> > There is a slight increase tried 3 times on Mac with similar conditions
> > Before:
> > Throughput: 999.362 plans/s, 300807.976 tasks/s
> > After:
> > Throughput: 1001.011 plans/s, 301304.197 tasks/s
> 
> The initial motivation for this change is memory over-use from our ShallowTrace data structure, is it possible to measure the memory usage before and after?

As you can minor GC is clean up more items before due to wrapper reference used Long
image

File: subprojects/parseq/src/main/java/com/linkedin/parseq/trace/ShallowTraceBuilder.java [hunk 2]
```diff
@@ -101,31 +110,58 @@ public ShallowTraceBuilder setResultType(final ResultType resultType) {
...
-      if (_endNanos == null) {
+      if (_endNanos == -1) {
         _endNanos = nanoTime;
       }
     }
     _resultType = resultType;
     return this;
   }
 
+  @Deprecated
```
Comment (mchen07) [inline subprojects/parseq/src/main/java/com/linkedin/parseq/trace/ShallowTraceBuilder.java:127]: As good practice. put javadoc 

`@deprecated replaced by ...`

File: subprojects/parseq/src/test/java/com/linkedin/parseq/trace/TestShallowTraceBuilder.java [hunk 3]
```diff
@@ -97,27 +104,56 @@ public void testEquals() {
...
-    builderCopy.setEndNanos(345L);
+    builderCopy.setNativeStartNanos(123L);
+    builderCopy.setNativePendingNanos(234L);
+    builderCopy.setNativeEndNanos(345L);
     builderCopy.setResultType(ResultType.ERROR);
     assertFalse(builder.build().equals(builderCopy.build()));
   }
+
+  @Test
+  public void testDeprecatedTimeMethods() {
```
Comment (mchen07) [inline subprojects/parseq/src/test/java/com/linkedin/parseq/trace/TestShallowTraceBuilder.java:132]: Let's cover several permutations for those fields we touched.
1. set and get all using deprecated methods
2. set using new method, and get using deprecated method
3. set using deprecated method, and get using new method
4. set and get all using new methods (which should already covered by all other tests you have fixed)

File: subprojects/parseq/src/main/java/com/linkedin/parseq/trace/ShallowTraceBuilder.java (hunk 2 above)
Comment (aman1309) [reply subprojects/parseq/src/main/java/com/linkedin/parseq/trace/ShallowTraceBuilder.java:127]: yup. missed here. added now

File: subprojects/parseq/src/test/java/com/linkedin/parseq/trace/TestShallowTraceBuilder.java (hunk 3 above)
Comment (aman1309) [reply subprojects/parseq/src/test/java/com/linkedin/parseq/trace/TestShallowTraceBuilder.java:132]: updated

File: subprojects/parseq/src/main/java/com/linkedin/parseq/trace/ShallowTraceBuilder.java [hunk 4]
```diff
@@ -124,6 +124,9 @@ public ShallowTraceBuilder setResultType(final ResultType resultType) {
     return this;
   }
 
+  /**
+   * Deprecated. Use {@link #setNativeStartNanos(long)} instead
```
Comment (mchen07) [inline subprojects/parseq/src/main/java/com/linkedin/parseq/trace/ShallowTraceBuilder.java:128]: we need to use @deprecated to generate proper javadoc for this.https://docs.oracle.com/javase/8/docs/technotes/guides/javadoc/deprecation/deprecation.html

**Pull request merged**